sdk/lib/core/string.dart

```diff
@@ -314,6 +314,9 @@ abstract class String implements Comparable<String>, Pattern {
   /// string.substring(1);    // 'artlang'
   /// string.substring(1, 4); // 'art'
   /// ```
+  ///
+  /// Both [start] and [end] must not be negative or greater than [length], and
+  /// [end], if provided, must be greater or equal to [start].
```

Comment:
I think I'd phrase it as "must be non-negative and no greater than [length].
Also end "must be greater than or equal to" start.

Good choice to document this!

Reply:
I applied your suggestion, thanks!